homeassistant/components/opower/config_flow.py

```diff
@@ -0,0 +1,119 @@
+"""Config flow for Opower integration."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+import logging
+from typing import Any
+
+from aiohttp.client_exceptions import ClientResponseError
+from opower import Opower, get_supported_utility_names
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers.aiohttp_client import async_create_clientsession
+
+from .const import CONF_UTILITY, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_UTILITY): vol.In(get_supported_utility_names()),
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+STEP_REAUTH_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+
+async def _validate_login(
+    hass: HomeAssistant, login_data: dict[str, str]
+) -> dict[str, str]:
+    """Validate login data and return any errors."""
+    api = Opower(
+        async_create_clientsession(hass),
+        login_data[CONF_UTILITY],
+        login_data[CONF_USERNAME],
+        login_data[CONF_PASSWORD],
+    )
+    errors = {}
+    try:
+        await api.async_login()
+    except ClientResponseError as err:
```

Comment:
I think it is preferred if you handle the clientresponse error in your api and have specific api exceptions that you raise. It gives you a lot more power and is better for error catching/ makes the code in core more readable

Reply:
Everybody understands HTTP status codes. I don't see much advantage of having my own exceptions that wrap them. In addition, the coordinator where you are supposed to do the polling already handles HTTP errors so I think it's fine to keep them as is.

Reply:
A bit late to the party, but fwiw I think I'm with Lash-L on this one. It wouldn't necessarily have to be a 1:1 wrapping of the status codes. A 400 can represent many different errors for example. The lib should just expose an interface representative of it's domain and http isn't part of that.

Reply:
Done.